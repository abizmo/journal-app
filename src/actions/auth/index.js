export { default as login } from './login';
export { default as loginWithEmail } from './loginWithEmail';
export { default as loginWithGoogle } from './loginWithGoogle';
export { default as logout } from './logout';
export { default as signOut } from './signOut';
export { default as signUp } from './signUp';

export const LOGIN = 'LOGIN';
export const LOGOUT = 'LOGOUT';

export const types = {
  LOGIN,
  LOGOUT,
};
