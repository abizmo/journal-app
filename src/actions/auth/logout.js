import { LOGOUT } from ".";

const logout = () => ({
  type: LOGOUT,
});

export default logout;
