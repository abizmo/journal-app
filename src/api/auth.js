import {
  createUserWithEmailAndPassword,
  onAuthStateChanged,
  signInWithEmailAndPassword,
  signInWithPopup,
  signOut as firebaseSignOut,
  updateProfile,
} from "firebase/auth";

import { auth, provider } from "../firebase";

export const onCurrentUserChanged = (cb) => onAuthStateChanged(auth, cb);
  
export const signIn = async ({ email, password }) => {
  const { user } = await signInWithEmailAndPassword(auth, email, password)

  return {
    name: user.displayName,
    userId: user.uid
  };
};

export const signInWithGoogle = () => signInWithPopup(auth, provider)
  .then(({ user }) => ({
    name: user.displayName,
    userId: user.uid
}));

export const signOut = () => firebaseSignOut(auth);

export const signUpWithEmailAndPassword = ({
  email,
  name,
  password
}) => createUserWithEmailAndPassword(auth, email, password)
  .then(({ user }) => updateProfile(user, { displayName: name })
  .then(() => ({ name, userId: user.uid})));
