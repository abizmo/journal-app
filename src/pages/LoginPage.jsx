import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Link } from 'react-router-dom';
import validator from 'validator';

import { loginWithEmail, loginWithGoogle } from '../actions/auth';
import { clearError, setError } from '../actions/ui';
import useForm from '../hooks/useForm';

const LoginPage = () => {
  const dispatch = useDispatch();
  
  const { error, loading } = useSelector(({ ui }) => ui);
  const [{ email, password }, handleInputChange ] = useForm({
    email: 'goccita@abizmo.dev',
    password: '123456',
  });
  
  const handleLogin = (evt) => {
    evt.preventDefault();
    if (isFormValid()) {
      dispatch(loginWithEmail({ email, password }));
    }
  };

  const isFormValid = () => {
    if (!validator.isEmail( email )) {
      dispatch(setError('Email is not valid'));
      return false;
    } if (password.length < 6) {
      dispatch(setError('Password is not valid or not match'));
      return false;
    }
    dispatch(clearError());
    return true;
  };

  const handleGoogleSignIn = () => {
    dispatch(loginWithGoogle());
  };

  return (
    <>
      <h3 className="auth__title mb-4">Login</h3>
      {
        error && (
          <div className="auth__alert-error">
            { error }
          </div>
        )
      }
      <form onSubmit={ handleLogin }>
        <input
          autoComplete="off"
          className="auth__input mb-2"
          name="email"
          onChange={handleInputChange}
          placeholder="tu@email.com"
          required
          type="email"
          value={email}
        />
        <input
          className="auth__input mb-2"
          name="password"
          onChange={handleInputChange}
          placeholder="TuContraseÃ±@"
          required
          type="password"
          value={password}
        />
        <button disabled={ loading }
          className="btn btn-block btn-primary mb-2"
          type="submit"
        >Login</button>
        <div className="auth__social-networks">
          <p className="auth__label">Login with social networks</p>
          <div className="google-btn mt-2" onClick={ handleGoogleSignIn }>
            <div className="google-icon-wrapper">
              <img
                alt="google button"
                className="google-icon"
                src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg"
              />
            </div>
            <p className="btn-text">
              <b>Sign in with google</b>
            </p>
          </div>
        </div>
        <Link
          className="link d-block"
          to="/auth/register"
        >Create new account</Link>
      </form>
    </>
  );
};

export default LoginPage;
