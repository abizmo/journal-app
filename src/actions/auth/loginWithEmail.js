import { login } from ".";

import { startAction, stopAction } from "../ui";
import { signIn } from "../../api/auth";
import { error } from "../../helpers/alert";

const loginWithEmail = ({ email, password }) => async (dispatch) => {
  try {
    dispatch(startAction());
    const user = await signIn({ email, password });
    dispatch(login(user));
  } catch ({ message }) {
    error(message);
  }
  dispatch(stopAction());
};

export default loginWithEmail;
