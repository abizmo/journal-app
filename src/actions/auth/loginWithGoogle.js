import { login } from ".";

import { signInWithGoogle } from "../../api/auth";
import { error } from "../../helpers/alert";

const loginWithGoogle = () => async (dispatch) => {
  try {
    const user = await signInWithGoogle();
    dispatch(login(user));
  } catch ({ message }) {
    error(message);
  }
};

export default loginWithGoogle;
