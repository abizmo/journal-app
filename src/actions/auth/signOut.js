import { logout } from '.';

import { clearNotes } from '../notes';
import { signOut as signOutApi } from '../../api/auth';
import { error } from "../../helpers/alert";

const signOut = () => (dispatch) => {
  signOutApi()
    .then(() => {
      dispatch(logout());
      dispatch(clearNotes());
    })
    .catch(({ message }) => error(message));
};

export default signOut;
