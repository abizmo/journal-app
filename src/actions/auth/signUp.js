import { login } from ".";

import { signUpWithEmailAndPassword } from "../../api/auth";
import { error } from "../../helpers/alert";

const signUp = (newUser) => async (dispatch) => {
  try {
    const user = await signUpWithEmailAndPassword(newUser);
    dispatch(login(user));
  } catch ({ message }) {
    error(message);
  }
};

export default signUp;
