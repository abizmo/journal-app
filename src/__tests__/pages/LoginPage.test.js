import { mount } from "enzyme";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router";
import createMockStore from "redux-mock-store";
import thunk from "redux-thunk";
import { loginWithEmail, loginWithGoogle } from "../../actions/auth";
import LoginPage from "../../pages/LoginPage";
jest.mock('../../actions/auth');

const middlewares = [thunk];
const mockStore = createMockStore(middlewares);
const initialState = {
  ui: {
    error: null,
    loading: false,
  },
};

describe('<LoginPage />', () => {
  let store = mockStore(initialState);
  store.dispatch = jest.fn();
  const wrapper = mount(
    <Provider store={store}>
      <MemoryRouter>
        <LoginPage />
      </MemoryRouter>
    </Provider>
  );

  beforeEach(() => {
    store.clearActions();
    jest.resetAllMocks();
  })

  test('should render right', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('should dispatch loginWithGoogle', () => {
    wrapper.find('.google-btn').simulate('click');

    expect(loginWithGoogle).toHaveBeenCalled();
  });
  
  test('should dispatch loginWithEmail with goccita@abizmo.dev', () => {
    const email = 'goccita@abizmo.dev';
    const password = '123456';
    wrapper.find('form').simulate('submit', { preventDefault: () => {} });

    expect(loginWithEmail).toHaveBeenCalledWith({ email, password });
  });
});
