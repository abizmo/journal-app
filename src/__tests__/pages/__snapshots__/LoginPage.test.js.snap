// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LoginPage /> should render right 1`] = `
Array [
  <h3
    className="auth__title mb-4"
  >
    Login
  </h3>,
  <form
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="auth__input mb-2"
      name="email"
      onChange={[Function]}
      placeholder="tu@email.com"
      required={true}
      type="email"
      value="goccita@abizmo.dev"
    />
    <input
      className="auth__input mb-2"
      name="password"
      onChange={[Function]}
      placeholder="TuContraseÃ±@"
      required={true}
      type="password"
      value="123456"
    />
    <button
      className="btn btn-block btn-primary mb-2"
      disabled={false}
      type="submit"
    >
      Login
    </button>
    <div
      className="auth__social-networks"
    >
      <p
        className="auth__label"
      >
        Login with social networks
      </p>
      <div
        className="google-btn mt-2"
        onClick={[Function]}
      >
        <div
          className="google-icon-wrapper"
        >
          <img
            alt="google button"
            className="google-icon"
            src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg"
          />
        </div>
        <p
          className="btn-text"
        >
          <b>
            Sign in with google
          </b>
        </p>
      </div>
    </div>
    <a
      className="link d-block"
      href="/auth/register"
      onClick={[Function]}
    >
      Create new account
    </a>
  </form>,
]
`;
